File: src/csharp/oop/VendingMachine/Vending/InventoryManager.cs
namespace VendingMachine.Vending;

public class InventoryManager
{
    private readonly Dictionary<string, Rack> _racks = new();

    public Rack GetRack(string rackName)
    {
        if (!_racks.TryGetValue(rackName, out var rack))
        {
            throw new ArgumentException($"Rack {rackName} does not exist");
        }

        return rack;
    }

    public bool AddProduct(string rackName, Product product, int count)
    {
        var rack = new Rack
        {
            Number = rackName,
            Product = product,
            Count = count
        };
        _racks.Add(rackName, rack);

        return true;
    }

    public Product DispenseProduct(string rackName)
    {
        if (!_racks.TryGetValue(rackName, out var rack))
        {
            throw new ArgumentException($"Rack {rackName} does not exist");
        }

        if (rack.Count <= 0)
        {
            throw new InvalidOperationException($"Rack {rackName} does not exist");
        }

        rack.Count--;

        return rack.Product;
    }
}

File: src/python/oop/VendingMachine/vending/inventory_manager.py
from .rack import Rack
from .product import Product


class InventoryManager:
    def __init__(self):
        self._racks_by_number = {}

    def get_rack(self, rack_number: str) -> Rack:
        if rack_number not in self._racks_by_number:
            raise ValueError(f"The '{rack_number}' not found")

        return self._racks_by_number[rack_number]

    def add_product_to_rack(self, rack_number: str, product: Product, count: int) -> bool:
        self._racks_by_number[rack_number] = Rack(rack_number, product, count)

        return True

    def dispense_product_from_rack(self, rack_number: str) -> Product:
        if rack_number not in self._racks_by_number:
            raise ValueError(f"The '{rack_number}' not found")

        rack = self._racks_by_number[rack_number]
        if rack.is_empty:
            raise Exception(f"The '{rack_number}' rack is empty")

        rack.count = rack.count - 1

        return rack.product
